src/network.rs

```diff
@@ -0,0 +1,495 @@
+			} => {
+				if endpoint.is_dialer() {
+					if let Some(sender) = self.pending_dials.remove(&peer_id) {
+						let _ = sender.send(Ok(()));
+					}
+				}
+				// this event is of interest,
+				// pass event to output event stream
+				self.event_sender
+					.send(Event::ConnectionEstablished { peer_id, endpoint })
+					.await
+					.expect("Event receiver not to be dropped.");
+			},
+			SwarmEvent::OutgoingConnectionError { peer_id, error } => {
+				if let Some(peer_id) = peer_id {
+					if let Some(sender) = self.pending_dials.remove(&peer_id) {
+						_ = sender.send(Err(error.into()));
+					}
+				}
+			},
+			SwarmEvent::Dialing(peer_id) => println!("Dialing {}", peer_id),
+			_ => {},
+		}
+	}
+
+	async fn handle_command(&mut self, command: Command) {
+		match command {
+			Command::StartListening { addr, sender } => {
+				_ = match self.swarm.listen_on(addr) {
+					Ok(_) => sender.send(Ok(())),
+					Err(e) => sender.send(Err(e.into())),
+				}
+			},
+			Command::AddAddress {
+				peer_id,
+				peer_addr,
+				sender,
+			} => {
+				self.swarm
+					.behaviour_mut()
+					.kademlia
+					.add_address(&peer_id, peer_addr.clone());
+				self.pending_kad_routing.insert(peer_id, sender);
+			},
+			Command::Dial {
+				peer_id,
+				peer_addr,
+				sender,
+			} => {
+				if let hash_map::Entry::Vacant(e) = self.pending_dials.entry(peer_id) {
+					match self
+						.swarm
+						.dial(peer_addr.with(Protocol::P2p(peer_id.into())))
+					{
+						Ok(()) => {
+							e.insert(sender);
+						},
+						Err(e) => {
+							_ = sender.send(Err(e.into()));
+						},
+					}
+				} else {
+					todo!("Implement logic for peer thats already beeing dialed.");
+				}
+			},
+			Command::Bootstrap { sender } => {
+				let query_id = self
+					.swarm
+					.behaviour_mut()
+					.kademlia
+					.bootstrap()
+					.expect("DHT not to be empty");
+
+				self.pending_kad_queries
+					.insert(query_id, QueryChannel::Bootstrap(sender));
+			},
+			Command::GetKadRecord {
+				key,
+				quorum,
+				sender,
+			} => {
+				let query_id = self.swarm.behaviour_mut().kademlia.get_record(key, quorum);
+				self.pending_kad_queries
+					.insert(query_id, QueryChannel::GetRecord(sender));
+			},
+			Command::PutKadRecord {
+				record,
+				quorum,
+				sender,
+			} => {
+				let query_id = self
+					.swarm
+					.behaviour_mut()
+					.kademlia
+					.put_record(record, quorum)
+					.expect("No put error.");
+
+				self.pending_kad_queries
+					.insert(query_id, QueryChannel::PutRecord(sender));
+			},
+		}
+	}
+}
+
+pub fn init(
+	seed: Option<u8>,
+	psk_path: &String,
+) -> Result<(Client, impl Stream<Item = Event>, EventLoop)> {
+	// Create a public/private key pair, either based on a seed or random
+	let id_keys = match seed {
+		Some(seed) => {
+			let mut bytes = [0u8; 32];
+			bytes[0] = seed;
+			let secret_key = ed25519::SecretKey::from_bytes(&mut bytes)
+				.expect("Error should only appear if length is wrong.");
```

Comment:
Instead of expect, we can use `?` and propagate error to caller. 

```
let secret_key = ed25519::SecretKey::from_bytes(&mut bytes)?;
```

Also, if we want to add additional context to error, we can use `context` fn before `?`.